src/NLog.Web.AspNetCore/NLogRequestLoggingMiddleware.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace NLog.Web
+{
+    /// <summary>
+    /// Middleware that writes all requests to Logger named "RequestLogging"
+    /// </summary>
+    /// <remarks>
+    /// - LogLevel.Error - Request failed with exception<br/>
+    /// - LogLevel.Warn  - Request completed with StatusCode >= 400<br/>
+    /// - LogLevel.Info  - Request completed normal StatusCode<br/>
+    /// </remarks>
+    public class NLogRequestLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly Microsoft.Extensions.Logging.ILogger _logger;
+
+        /// <summary>
+        /// Initializes new instance of the <see cref="NLogRequestLoggingMiddleware"/> class
+        /// </summary>
+        /// <remarks>
+        /// Use the following in Startup.cs:
+        /// <code>
+        /// public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        /// {
+        ///    app.UseMiddleware&lt;NLog.Web.RequestLoggingMiddleware&gt;();
+        /// }
+        /// </code>
+        /// </remarks>
+        public NLogRequestLoggingMiddleware(RequestDelegate next, ILoggerFactory loggerFactory)
+        {
+            _next = next;
+            _logger = loggerFactory?.CreateLogger("RequestLogging") ?? Microsoft.Extensions.Logging.Abstractions.NullLogger.Instance;
+        }
+
+        /// <summary>
+        /// Executes the middleware.
+        /// </summary>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+                LogHttpRequest(httpContext, null);
+            }
+            catch (Exception exception) when (LogHttpRequest(httpContext, exception))
+            {
+                // Logging complete
+            }
+        }
+
+        private bool LogHttpRequest(HttpContext httpContext, Exception exception)
+        {
+            if (exception != null)
+            {
+                _logger.LogError(exception, "HttpRequest Exception");
+            }
+            else
+            {
+                var statusCode = httpContext.Response?.StatusCode ?? 400;
+                if (statusCode >= 400)
+                {
+                    _logger.LogWarning("HttpRequest Failed");
+                }
+                else
+                {
+                    _logger.LogInformation("HttpRequest Completed");
+                }
+            }
+
+            return false;
```

Comment:
What is this return value saying?

Reply:
Says that the first-chance-exception-filter has not handled the exception, so it should just continue as a normal exception.